gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/FirestoreDataStore.java

```diff
@@ -91,30 +97,37 @@ public Completable applyMutations(ImmutableCollection<Mutation> mutations, User
   private Task<?> applyMutationsInternal(ImmutableCollection<Mutation> mutations, User user) {
     WriteBatch batch = db.batch();
     for (Mutation mutation : mutations) {
-      addMutationToBatch(mutation, user, batch);
+      try {
+        addMutationToBatch(mutation, user, batch);
+      } catch (DataStoreException e) {
```

Comment:
Should we use exceptions for this case? 

I'm of the opinion that exceptions are more useful for states that are actually unrecoverable/catastrophic--places where crashing w/ a message + trace is a valid course of action, whereas when we do stuff like ignore or log a potential value we don't care about and otherwise continue running some kind of value wrapper type is more appropriate—it doesn't make much of a difference at the end of the day, so I would bother if you have a different stance on this.

Reply:
If exceptions were unrecoverable, then there'd be no need for try/catch. Sun says "An exception is an event that occurs during the execution of a program that disrupts the normal flow of instructions." Issues like these with bad input data prevent the deserialization of remote entities, but don't break the entire app. Where we handle these and how obviously an art and not a science, and I agree sometimes it may make sense to use value wrappers instead of exceptions. But for values that really really need to be there for the data be used (i.e., we're dropping data, the database is corrupt) an exception definitely seems like the right mechanism. 

@dturner in case you have anything to add!

Reply:
That sounds reasonable to me. I'm fine with using them--I think at times they compete with alternatives like returning Result(x)--thanks for the explanation!

But i will drop [this chart](https://www.progsbase.com/blog/flow-charts-of-programming-language-constructs/#11.-exception) :)

Reply:
+1, especially when we need a was to propagate errors in streams.

I haven't seen that before, but at first glance it seems they may be confusing code readability with graph complexity, which I intuitively believe may not be strictly correlated. The abstraction provided by exception propagation could allow you to represent the call stack as a single box, in some sense... but I could be wrong.

Reply:
That's a good point! I guess I just think things are easier to reason about the less we mix styles--but it probably is simpler to go with exceptions in certain cases--I just hope doing things on a case-by-case basis doesn't make it more difficult to alter things in the long run. But I think our choices/boundaries are pretty clear right now, so it probably won't be an issue.